mbbsd: add email_challenge

chhsiao1981 opened the pull request:
mbbsd: add email_challenge

Currently we verify email again only in AccountRecovery.
We should verify email as well when the user intends to reset the password.

It is possible that we will need to verify the email when changing
the contact-email in the future.
Use email_challenge instead of passwd_challenge for this generalization.

The implementation is based on AccountRecovery::EmailCodeChallenge.

It is better to review #113 before merging this PR,
in that #113 enables us differentiating register-email vs. contact-email.

wens commented on mbbsd/email_challenge.cc:
```diff
@@ -0,0 +1,203 @@
+  if (email_matches) {
+    // We only want to send email if the user input the matching address.
+    // Silently fail if email is not matching, so that user cannot guess other
+    // user's email address.
```
I'm not sure what you want to achieve with this. The user's email is already visible through the user info menu, even the email stored in the verification database.

chhsiao1981 replied:
I wrote email_challenge.cc based on recovery.cc

This is just trying to minimize the difference between recovery.cc and email_challenge.cc:

https://github.com/ptt/pttbbs/blob/master/mbbsd/recover.cc#L227

chhsiao1981 replied:
`all_emails` includes the contact email and the "original verified email".

It is possible that some user's contact email got updated by some bad guy, 
but the user can still use the "original verified email" to do the account-related operations (retrieve the account / change the contact-email again.)

chhsiao1981 replied:
`all_emails` includes "contact email" and "original verified email".

It's possible that some user's contact email got updated by some bad guy,
but the user can still use "original verified email" to do account-related operations (retrieve the account / change the email.)

wens replied:
What I meant was the original verified email is visible via (U)ser -> (I)nfo -> (V), so the "prevent bad user from guessing" is moot.

Plus you already copied the implementation. What's the point of minimizing the difference? It's not like you are reusing it.

chhsiao1981 replied:
I'm sorry that I was not clear earlier～

1. The purpose of "email-challenge" is to send the challenge to the email, not just "guessing the verified email".
    From my understanding, now we have "contact-email" and "verified-email".
    There is only 1 "contact-email" and currently it can be easily modified.
    "verified-email" is harder to be modified because it requires valid verifiable email-address.

    It is possible that some user's contact-email got changed by the bad guy,
    but the user still can access the verified email.
    the user would like to recover the password / contact email through the verified email in this case.

2. I did want to merge the two functions. After all it should be the single function doing email-challenge.
    However, I'm not sure whether it is ok to change `AccountRecovery::EmailCodeChallenge` as well in this PR.
    So I just leave `AccountRecovery::EmailCodeChallenge` intact in this PR.

    I am very happy to merge the two functions in this PR if it is ok to do so.

wens replied:
Re: 1.  But this is a separate function, i.e. 2FA _before_ changing password / email, so the user (legit or otherwise) already has access to said info. Why bother having the user input it again? Just send the challenge email to both the verified email and contact email.

Re: 2. Either merge them (which I doubt would work or look pretty), or just write functions that suite your need. Duplicating it is the worst choice.

chhsiao1981 replied:
@wens 

Thanks for the reply.

1. I didn't think of this alternative setup. I'll update to this setup.
2. I'll merge these two functions to simplify the code.

chhsiao1981 commented:
This PR requires email-challenge only during "resetting password".

Therefore, we don't need to require resetting contact-email before approving this PR.

The potential requiring email-challenge when changing contact-email will be in other PR
(after approving this PR.)

chhsiao1981 commented:
hi @wens ～

I believe that I followed your guideline and finished updating this PR.

USE_2FA with resetting-password / changing contact-email is implemented
in this PR as well.

Can you take a look again？～

Thanks!

wens commented on mbbsd/register.c:
```diff
@@ -1743,15 +1745,29 @@ static int notify_email_change(const char *userid, const char *email)
+change_contact_email(bool skip_same_email_check, bool two_factor_authentication)
 {
     char email[EMAILSZ] = {};
     memcpy(email, cuser.email, sizeof(email));
```
I think we should force a reload from disk, i.e. call `pwcuReload()`, before this, in case another login instance changed the email. This should be a new precursor commit.

wens commented on include/email_challenge.hpp:
```diff
@@ -0,0 +1,9 @@
+
+namespace emailchallenge {
+
+bool EmailChallenge(bool check_input_email, const std::string &input_email, const userhandle::UserHandle &user, const int y, const std::string &prompt, const std::string &ip, const std::string &filename, int *out_y);
```
Could you wrap the line at 80 or 100 characters, and align the following lines with the open parentheses?

wens commented on include/proto.h:
```diff
@@ -228,6 +228,9 @@ void auto_backup(void);
 const char *ask_tmpbuf(int y);
 
+/* email-challenge */
+int email_challenge(const char *email, const userec_t *user, const int y, const char *prompt, const char *ip, const char *filename, int *out_y);
```
Same here.

wens reviewed (commented):
Can you split this into two commits? The first commit only moves code and adds the namespace. The second one modifies the code to add whatever you added.

This would make it much more easier to diff and review.

chhsiao1981 commented:
@wens 

Thanks for the comments!
I'll update it asap～

chhsiao1981 commented:
hi @wens ～

Thank you so much for the comments~
I believe that I followed your guideline and finished updating this PR.

Can you take a look again？～

Thanks!

wens commented:
Please move [mbbsd: resync cuser.email in the beginning of change_contact_email](https://github.com/ptt/pttbbs/pull/114/commits/a9c153b5d30203338203404d9df5ac7a222d1935) to be the first, i.e. oldest, commit in the PR. Also please drop the period at the end.

wens commented on mbbsd/email_challenge.cc:
```diff
@@ -0,0 +1,270 @@
+#include "user_handle.hpp"
+#include "email_challenge.hpp"
+
+namespace emailchallenge {
```
Please add underscore between the two words. That makes it easier to read.

wens commented on include/proto.h:
```diff
@@ -580,7 +589,7 @@ void check_register(void);
     char *email, const char **errmsg, const char **notice_file);
 void register_mail_complete_and_exit();
-int change_contact_email(bool skip_same_email_check);
+int change_contact_email(bool skip_same_email_check, bool two_factor_authentication);
```
Better name for second parameter: `require_two_factor_auth`

wens commented on mbbsd/email_challenge.cc:
```diff
@@ -3,6 +3,12 @@ extern "C" {
 #include "daemons.h"
 }
 
+#ifdef USE_2FA
```
Call this `USE_EMAIL_2FA` or something, in case later we have other forms of 2FA.

wens commented on mbbsd/email_challenge.cc:
```diff
@@ -3,6 +3,12 @@ extern "C" {
 
+#ifdef USE_2FA
+# ifndef USE_VERIFYDB_ACCOUNT_RECOVERY
+#   error "USE_2FA requires USE_VERIFYDB_ACCOUNT_RECOVERY and USE_VERIFYDB"
```
VERIFYDB is not strictly needed. You can always skip `LoadVerifyDbEmail()` and rely solely on user's listed email.

wens commented on mbbsd/user_handle.cc:
```diff
@@ -0,0 +1,34 @@
+#include <string>
+#include "user_handle.hpp"
+
+namespace userhandle {
```
Add an underscore between the words.

wens commented on mbbsd/email_challenge.cc:
```diff
@@ -0,0 +1,270 @@
+                              const std::string &ip,
+                              const std::string &filename);
+static std::string GenCode(size_t len);
+static void UserErrorExit();
```
No need to forward declare static stuff. Just order them properly.

wens commented:
> Can you split this into two commits? The first commit only moves code and adds the namespace.
You should keep changes to a minimal. That includes any changes to comments and stuff. It should be a pure code move, with only necessary changes to make it compile.

If you use `git show <commit>--color-moved --color-moved-ws=ignore-all-space`, it should mark most of the commit as moved, and only bits where you added namespace as deletions and additions. You could even forgo the namespace and just have it in the global namespace.

wens commented:
Simply put, if you need to make modifications so the code fits, do them in separate commits. A code move is just a code move, there should not be any other changes. One commit equals one logical commit, without breaking the build.

chhsiao1981 commented:
hi @wens ～

Thank you so much for your comments!
I'll update accordingly asap～

chhsiao1981 committed:
mbbsd: resync cuser.email in the beginning of change_contact_email

Now we want to email / verify the existing cuser.email in change_contact_email.
We would like to re-sync pwcu to get the most updated cuser.email
before emailing / verifying.

chhsiao1981 committed:
mbbsd: refactor ::AccountRecovery::EmailCodeChallenge to EmailCodeChallenge

Currently we verify email (contact-email or register-email as valid email) again
only in AccountRecovery.

Refactor ::AccountRecovery::EmailCodeChallenge to EmailCodeChallenge
and the corresponding helper static functions to prepare for the extension
to verify email in other sensitive scenarios.

chhsiao1981 committed:
mbbsd: renaming email-challenge related code

Renaming email-challenge related code to be more semantically correct.

Removing tail-underscore in the parameters and changing SendRecoveryCode
to SendChallengeCode.

chhsiao1981 committed:
mbbsd: add email_challenge as 2FA

Currently we verify email (contact-email or register-email as valid email)
again only in AccountRecovery. We should verify email as well
when the user intends to reset the password or change the contact email.

This feature is enabled with USE_EMAIL_2FA defined in pttbbs.conf.

chhsiao1981 commented:
hi @wens ～

Thank you so much for the comments~
I believe that I followed your guideline and finished updating this PR.

Can you take a look again？～

Thanks!

chhsiao1981 commented:
closed for now due to stalled for a very long time.
